### DISCUSSION TITLE
LWM2M Firmware Package URI

### DISCUSSION THREAD

**Discussion by thomasBisseling** [zephyrproject-rtos/zephyr#61024] [Q&A] [answered]
What does Zephyr with the Firmware Package URI? In the source code it looks like it does a lot. But does it also write it to the flash for the FOTA? Please explain what it exactly does for you

**Comment by RomainPelletant** [comment]
Hi Thomas,

It implements the mechanism (state machine and download) for this OTA pull method. It comes to you to write each block received into the memory (second partition) to update the firmware.
To do it, you have to implement callbacks  like done in the LwM2M sample here : https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/lwm2m_client/src/firmware_update.c

According to LwM2M specifications, you shall set URI (ID 1) and execute update (ID2) once downloaded (means once fully written in memory and ask bootloader to switch and reboot)

Only CoAP download is supported as suggested in specifications to avoid increasing fw size with TCP stack for HTTP.

**Reply by thomasBisseling** [reply]
Only when i am looking into the code. It looks like Zephyr is already downloading the package using the URI. Is that right? So how can i access the package data

**Reply by RomainPelletant** [reply, marked as answer]
Download is done by block. In the code sample provided, you have to write each block ( received in static int firmware_block_received_cb(uint16_t obj_inst_id,...) into the memory.

So each new block fw received could be written in memory using flash API or 
DFU API (https://docs.zephyrproject.org/latest/services/device_mgmt/dfu.html)
